packages/chain-controller/src/ChainController.test.ts

```diff
@@ -0,0 +1,133 @@
+import type { InternalAccount } from '@metamask/keyring-api';
+import type { SnapController } from '@metamask/snaps-controllers';
+import type { SnapId } from '@metamask/snaps-sdk';
+import type { SnapRpcHookArgs } from '@metamask/snaps-utils';
+
+import type { ChainControllerMessenger } from './ChainController';
+import { ChainController } from './ChainController';
+
+describe('ChainController', () => {
+  const snapId = 'local:localhost:3000' as SnapId;
+  const snapClient = {
+    submitRequest: jest.fn(),
+  };
+  const snapController = {
+    handleRequest: async (request: { snapId: SnapId }) => {
+      if (request.snapId === snapId) {
+        return snapClient.submitRequest(request);
+      }
+      throw new Error(`Unknown Snap: ${request.snapId as string}`);
+    },
+  } as unknown as SnapController;
+
+  const messenger = {
+    call: jest.fn(
+      async (
+        action: string,
+        request: SnapRpcHookArgs & { snapId: SnapId },
+      ): Promise<unknown> => {
+        switch (action) {
+          case `SnapController:handleRequest`:
+            return await snapController.handleRequest(request);
+
+          default:
+            throw new Error(`Unknown action: ${action}`);
+        }
+      },
+    ),
+    registerActionHandler: jest.fn(),
+    registerInitialEventPayload: jest.fn(),
+    publish: jest.fn(),
+    subscribe: jest.fn(),
+  } as unknown as ChainControllerMessenger;
+
+  const makeController = () => {
+    return new ChainController({
+      state: {},
+      messenger,
+    });
+  };
+
+  const address = 'bc1qrp0yzgkf8rawkuvdlhnjfj2fnjwm0m8727kgah';
+  const scope = 'bip122:000000000019d6689c085ae165831e93';
+  const asset = `${scope}/asset:0`;
+
+  describe('registerProvider', () => {
+    it('returns false if there is no known provider', () => {
+      const controller = makeController();
+
+      expect(controller.hasProviderFor(scope)).toBe(false);
+    });
+
+    it('registers a chain provider for a given scope', () => {
+      const controller = makeController();
+
+      expect(controller.registerProvider(scope, snapId)).toBeDefined();
+      expect(controller.hasProviderFor(scope)).toBe(true);
+    });
+
+    it('fails to register another provider for an existing scope', () => {
+      const controller = makeController();
+      const anotherSnapId = 'local:localhost:4000' as SnapId;
```

Comment:
I must say, that I don't know... Since it's an opaque type (and I haven't seen any "factory"/"builder" methods to make one out of a `string`, I do use type assertion like here: https://github.com/MetaMask/snaps/blob/main/packages/snaps-controllers/src/snaps/SnapController.test.tsx#L5528-L5529

Reply:
Ah, I see, interesting. Perhaps this is the only way, but I've asked the Snaps team about this just in case there is a better way.